sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs

```diff
@@ -0,0 +1,339 @@
+            }
+        }
+
+        private static void CheckMemberName(MemberDeclarationSyntax member, SyntaxToken identifier,
+            SyntaxNodeAnalysisContext context)
+        {
+            var symbol = context.SemanticModel.GetDeclaredSymbol(member);
+            if (symbol == null)
+            {
+                return;
+            }
+
+            if (IsTypeComRelated(symbol.ContainingType) ||
+                symbol.GetInterfaceMember() != null ||
+                symbol.GetOverriddenMember() != null ||
+                symbol.IsExtern)
+            {
+                return;
+            }
+
+            if (identifier.ValueText.StartsWith("_", StringComparison.Ordinal) ||
+                identifier.ValueText.EndsWith("_", StringComparison.Ordinal))
+            {
+                context.ReportDiagnostic(Diagnostic.Create(rule_MethodName, identifier.GetLocation(), TypeKindNameMapping[member.Kind()],
+                    identifier.ValueText, MessageFormatUnderscore));
+                return;
+            }
+
+            if (identifier.ValueText.Contains("_"))
+            {
+                return;
+            }
+
+            string suggestion;
+            if (!IsMemberNameValid(identifier.ValueText, out suggestion))
+            {
+                var messageEnding = string.Format(MessageFormatNonUnderscore, suggestion);
+                context.ReportDiagnostic(Diagnostic.Create(rule_MethodName, identifier.GetLocation(), TypeKindNameMapping[member.Kind()],
+                    identifier.ValueText, messageEnding));
+            }
+        }
+
+        private static bool IsMemberNameValid(string identifierName, out string suggestion)
+        {
+            if (identifierName.Length == 1)
+            {
+                suggestion = identifierName.ToUpperInvariant();
+                return suggestion == identifierName;
+            }
+
+            var suggestionBuilder = new StringBuilder(identifierName.Length);
+
+            foreach (var part in SplitToCamelCase(identifierName))
+            {
+                suggestionBuilder.Append(SuggestLessUppercase(part, 1));
+            }
+
+            suggestionBuilder[0] = char.ToUpperInvariant(suggestionBuilder[0]);
+            suggestion = SuggestCapitalLetterAfterNonLetter(suggestionBuilder);
+
+            return suggestion == identifierName;
+        }
+
+        private static bool IsTypeNameValid(string identifierName, bool requireInitialI, bool allowInitialI,
+            bool areUnderscoresAllowed, out string suggestion)
+        {
+            if (identifierName.Length == 1)
+            {
+                suggestion = identifierName.ToUpperInvariant();
+                return suggestion == identifierName;
+            }
+
+            var idealNameVariant = new StringBuilder(identifierName.Length);
+            var acceptableNameVariant = new StringBuilder(identifierName.Length);
+
+            var parts = SplitToCamelCase(identifierName).ToList();
```

Comment:
These are not words. It could also be a symbol, or a two-letter acronym, which makes it a bit confusing.